src/jit/compiler.cpp

```diff
@@ -2440,147 +2484,111 @@ void Compiler::compSetProcessor()
+                                }
+                            }
+                        }
+                    }
                 }
             }
-            if (jitFlags.IsSet(JitFlags::JIT_FLAG_USE_SSE42))
+
+            if (configEnableIsa(jitFlags, InstructionSet_LZCNT))
             {
-                if (configEnableISA(InstructionSet_SSE42))
-                {
-                    opts.setSupportedISA(InstructionSet_SSE42);
+                opts.setSupportedISA(InstructionSet_LZCNT);
 #ifdef _TARGET_AMD64_
-                    opts.setSupportedISA(InstructionSet_SSE42_X64);
-#endif
-                }
-            }
-            if (jitFlags.IsSet(JitFlags::JIT_FLAG_USE_SSSE3))
-            {
-                if (configEnableISA(InstructionSet_SSSE3))
-                {
-                    opts.setSupportedISA(InstructionSet_SSSE3);
-                }
-            }
-            // AES and PCLMULQDQ requires 0x660F38/A encoding that is
-            // only used by SSSE3 and above ISAs
-            if (jitFlags.IsSet(JitFlags::JIT_FLAG_USE_AES))
-            {
-                if (configEnableISA(InstructionSet_AES))
-                {
-                    opts.setSupportedISA(InstructionSet_AES);
-                }
-            }
-            if (jitFlags.IsSet(JitFlags::JIT_FLAG_USE_PCLMULQDQ))
-            {
-                if (configEnableISA(InstructionSet_PCLMULQDQ))
-                {
-                    opts.setSupportedISA(InstructionSet_PCLMULQDQ);
-                }
+                opts.setSupportedISA(InstructionSet_LZCNT_X64);
+#endif // _TARGET_AMD64_
             }
-        }
 
-        // There are currently two sets of flags that control instruction sets that require the VEX encoding:
-        // These are the general EnableAVX flag and the individual ISA flags. We need to
-        // check both for any given isa.
-        if (JitConfig.EnableAVX())
-        {
-            if (jitFlags.IsSet(JitFlags::JIT_FLAG_USE_AVX))
-            {
-                if (configEnableISA(InstructionSet_AVX))
-                {
-                    opts.setSupportedISA(InstructionSet_AVX);
-                }
-            }
-            if (jitFlags.IsSet(JitFlags::JIT_FLAG_USE_FMA))
-            {
-                if (configEnableISA(InstructionSet_FMA))
-                {
-                    opts.setSupportedISA(InstructionSet_FMA);
-                }
-            }
-            if (jitFlags.IsSet(JitFlags::JIT_FLAG_USE_AVX2))
-            {
-                if (configEnableISA(InstructionSet_AVX2))
-                {
-                    opts.setSupportedISA(InstructionSet_AVX2);
-                }
-            }
-            if (jitFlags.IsSet(JitFlags::JIT_FLAG_USE_BMI1))
+            // We currently need to also check that AVX is supported as that controls the support for the VEX encoding in the emitter.
+            if (configEnableIsa(jitFlags, InstructionSet_BMI1) && compSupports(InstructionSet_AVX))
```

Comment:
I want to, in a follow-up PR, change `canUseVexEncoding()` to be controlled by an independent (debug only) flag: `COMPlus_EnableVEX`. In 2.2 and prior, this is controlled by `COMPlus_EnableAVX`, but from a functional perspective, this only controlled whether `S.N.Vector<T>` was 32-bytes or 16-bytes.

@CarolEidt, thoughts here?

Reply:
The thought process is that:
* We need to support the VEX encoding regardless of the value of `EnableAVX` due to `BMI1` and `BMI2`
* Users will likely want to emit the more efficient VEX encoded instructions even if they also want `S.N.Vector<T>` to be 16-bytes
* There should be no functional difference between `EnableVEX=0` and `EnableVEX=1` (only differences in the encoding generated), so it is likely not worthwhile for the flag to be available in retail builds
  * This differs from the other ISAs, where we need to provide the ability for users to test the difference code-paths without modifying their source or binaries

Reply:
> We need to support the VEX encoding regardless of the value of EnableAVX due to BMI1 and BMI2
> There should be no functional difference between EnableVEX=0 and EnableVEX=1

In my personal opinion, it is not necessary to add a new knob `EnableVEX` for such strange situations unless anyone actually needs it in real application testing. I doubt how many people understand VEX encoding :)

> Users will likely want to emit the more efficient VEX encoded instructions even if they also want S.N.Vector<T> to be 16-bytes

Currently, 16-byte `S.N.Vector<T>` is VEX-encoded on AVX-only machine (`EnableAVX2=0`).

Reply:
> In my personal opinion, it is not necessary to add a new knob EnableVEX for such strange situations unless anyone actually needs it in real application testing

Right, this is why I said it would be a debug only knob, since the JIT still needs it to validate we don't regress the non-VEX code paths (but since it shouldn't be important for anyone else).

> Currently, 16-byte S.N.Vector<T> is VEX-encoded on AVX-only machine (EnableAVX2=0).

Right, but this was a recent change that came out of the HWIntrinsic work (it wasn't the case for 2.2 and prior). With 3.0, if the user currently disables `AVX` or `AVX2`, then `S.N.Vector<T>` will be 16-bytes and if they only disable `AVX2`, then `S.N.Vector<T>` will still be VEX encoded. I was saying that we should also emit the VEX encoding if the user disables the `AVX` HWIntrinsics, as that generally a more efficient encoding/etc.

We also have `INTERNAL_SIMD16ByteOnly`, which disables AVX2 in order to force `S.N.Vector<T>` to be 16-bytes and that needs to be reasoned about as well (it is probably more beneficial for this to change the size of `S.N.Vector<T>` without disabling `AVX`/`AVX2`, for example).

Reply:
> but this was a recent change that came out of the HWIntrinsic work (it wasn't the case for 2.2 and prior).

Hmm, I think that is also available on 2.1 if they use an AVX-only machine (but no release knob to control).

Reply:
Ah, right. It would have worked in 2.1 since we had changed `useVexEncoding` for the hwintrinsics preview. But only for the few cups with only avx support

Reply:
> Users will likely want to emit the more efficient VEX encoded instructions even if they also want S.N.Vector<T> to be 16-bytes

I **really** don't think we want to encourage users to use `COMPlus_EnableAVX=0` as a way to set the length of vectors. It also needs to imply no VEX encoding (there is/was at least one firmware bug that used this in the registry as a workaround, and I think the encoding implication is required). If we really want to expose that, then we need another flag. But I'm not in favor of that without a compelling argument.

Reply:
> I really don't think we want to encourage users to use COMPlus_EnableAVX=0 as a way to set the length of vectors

This is the only way users can control that behavior today, as there isn't any other retail flag that allows that behavior. I agree that a separate flag would be the better long-term option.

> It also needs to imply no VEX encoding (there is/was at least one firmware bug that used this in the registry as a workaround, and I think the encoding implication is required). If we really want to expose that, then we need another flag. But I'm not in favor of that without a compelling argument.

The argument is that disabling `AVX` will also disable `BMI1` and `BMI2` (which also use the VEX encoding), even though these are not in the same ISA hierarchy.
* Even though we will likely never encounter a CPU that has BMI1/BMI2 support without AVX support; it is possible for `AVX.IsSupported` to return false while `BMI1/BMI2.IsSupported` returns true, because AVX has an additional requirement that the OS supports the YMM register state, while BMI1/BMI2 do not (and it is possible to dynamically enable/disable YMM state support in Windows by calling `bcdedit` and rebooting).